File: Shop/Python/shop.py
"""
Olga Rozhdestvina
GMIT MULTI-PARADIGM-PROGRAMMING 2020
Shop Simulator in Python 
Procedural Programming
"""

from dataclasses import dataclass, field    # for procedual programming
import shutil                               # To move a temporary file into stock.csv               
from typing import List                     # For working with shopping lists and shop stock
import csv                                  # For process CSVs
import subprocess                           # For clearing the screen
from tempfile import NamedTemporaryFile     # To create a temporary file when writing a CSV


filename = '../stock.csv'
# Clear the screen
subprocess.call("cls", shell=True)
print("\t\t Welcome to Olga's Shop!\n++++++++++++++++++++++++++++++++++++++++++++++++++++\n")


@dataclass
class Product:
    name: str
    price: float = 0.0

@dataclass
class ProductStock:
    product: Product
    quantity: int

@dataclass
class Shop:
    cash: float = 0.0
    stock: List[ProductStock] = field(default_factory=list)

@dataclass
class Customer:
    name: str = ""
    budget: float = 0.0
    shopping_list: List[ProductStock] = field(default_factory=list)


# PART 1: SHOP

# Stock the shop from CSV using the above classes
def create_and_stock_shop():
    s = Shop()
    # Open the stock.csv file
    try:
        with open(filename) as csv_file:
            csv_reader = csv.reader(csv_file, delimiter=',')
            # Separate the info in the first row
            first_row = next(csv_reader)
            s.cash = float(first_row[0])
            # Stock up the rest of the csv as product list
            for row in csv_reader:
                p = Product(row[0], float(row[1]))
                ps = ProductStock(p, int(row[2]))
                # Add to the list of items
                s.stock.append(ps)
        return s
    except FileNotFoundError:
        print(f">>>> ERROR: cannot open {filename}")

# Display the shop
def print_shop(s):
    print(" *********************************************\n")
    print(f"\tCASH AT THE SHOP: €{s.cash:.2f}\n")
    print(" *********************************************\n")
    print(" Product\t\tPrice\t    Qty")
    print(" _____________________________________________\n")
    for i in s.stock:
        print(f" {i.product.name}")
        print(f"\t\t\t {i.product.price}\t    {i.quantity}")
        print("---------------------------------------------\n")

# Call shop from the menu
def view_shop():
    # Clear the screen.
    subprocess.call("cls", shell=True)
    s = create_and_stock_shop()
    print_shop(s)




# PART 2: LIVE MODE

"""
Function that collects data from a customer
and displays shop's available products to choose from.
"""
def place_order():
    # Clear the screen.
    subprocess.call("cls", shell=True)
    s = create_and_stock_shop()
    c = Customer()

    # Ask customer to enter his/her details
    c.name = input(" Please enter your name: ").strip().capitalize()
    while True:
        try:
            c.budget = float(input(" Please enter your budget: "))
            break
        except:
            print(" >>>> ERROR: Please enter a number")
            continue
    print(f"\n\t Hello, {c.name}!")
    print(" +++++++++++++++++++++++++")
    print("\n Choose from our products:\n")
    for i in s.stock:
        print(f" {i.product.name}")  

    shopping_list(s, c)    
    customers_order(c)


# Creating a shopping list from customer's input
def shopping_list(s, c):
    # Ask a customer to enter products and their quantity
    continue_order = 0
    while continue_order != 'n':
        product = input("\n Please enter product: ").strip().lower()
        try: 
            quantity = int(input(" Please enter quantity: "))
        except ValueError:
            print(" >>>> ERROR: please enter a number")
            continue

        # Create a shopping list from the products the customer entered
        for i in s.stock:
            if product == i.product.name.lower():
                total_cost = quantity * i.product.price
                print(f"\n >>>> The cost of it is €{total_cost:.2f}.")
                add_to_basket = input("\n Would you like to add it to your basket? [Y/N]: ").strip().lower()
                if add_to_basket == 'y':
                    p = i.product
                    ps = ProductStock(p, quantity)
                    c.shopping_list.append(ps)
                continue_order = input("\n Would you like to order something else? [Y/N]: ").strip().lower()
                print("-------------------------------------------------")    


# PART 3: READ CUSTOMER ORDER FROM CSV

# Choose between 3 CVS files to read
def import_order():
    # Clear the screen.
    subprocess.call("cls", shell=True)
    print("-----------------------------------------------------")
    print("\n Please choose the file you would like to import.\n")
    print("\n PRESS\n\t-[1] for Customer 1\n\t-[2] for Customer 2\n\t-[3] for Customer 3\n\t- or any other key to Return to the menu")
    answer = input("\n Your choice: ").strip()
    # normal order
    if answer == '1':
        c = read_customer('../customer1.csv')
        customers_order(c)
    # not enough money
    elif answer == '2':
        c = read_customer('../customer2.csv')
        customers_order(c)
    # not enough stock
    elif answer == '3':
        c = read_customer('../customer3.csv')
        customers_order(c)


# Load a customer's CSV file
def read_customer(file_path):
    try:
        with open(file_path) as csv_file:
            csv_reader = csv.reader(csv_file, delimiter=',')
            # Separate the info in the first row
            first_row = next(csv_reader)
            c = Customer(first_row[0], float(first_row[1]))
            # Create a shopping list for a customer
            for row in csv_reader:
                name = row[0]
                quantity = int(row[1])
                # Check the shop for the items that are in customer's CSV
                s = create_and_stock_shop()
                for i in s.stock:
                    # if product in the customer's order, add to the shopping list(ignore capitalisation)
                    if i.product.name.lower() == name.lower():
                        p = i.product
                        ps = ProductStock(p, quantity)
                        c.shopping_list.append(ps)           
            return c
    except FileNotFoundError:
        print(f" >>>> ERROR: cannot open {file_path}")


"""
Function that displays customer's credentials,
his structured shopping list,
calculates total cost of the order,
checks if the customer has enough funds to pay for the transaction
"""

def customers_order(c):
    total_cost = 0.0
    # Clear the screen.
    subprocess.call("cls", shell=True)
    print(" *********************************************\n")
    print(f" CUSTOMER: {c.name}\n BUDGET:   €{c.budget:.2f}\n")
    print("*********************************************\n")
    print(" Product\t\tPrice\tQty\tCost\n")
    print(" _____________________________________________\n")
    
    # print the details of each product in the shop
    s = create_and_stock_shop()
    for i in c.shopping_list:
        cost = i.quantity * i.product.price
        if i.quantity > 10:
            print(f" Discounted\n {i.product.name}")
            discounted = cost - (cost * 0.1)
            print(f"\t\t\t {i.product.price}\t {i.quantity} \t{discounted:.2f}")
            print("---------------------------------------------\n")
            # add the line cost to the total
            total_cost += discounted
        else:
            print(f" {i.product.name}")
            print(f"\t\t\t {i.product.price}\t {i.quantity} \t{cost:.2f}")
            print("---------------------------------------------\n")
            # add the line cost to the total
            total_cost += cost
    print(f"\n Total\t\t\t\t\t{total_cost:.2f}")
    print("---------------------------------------------\n")
    process_or_not = input(f"\n Would you like to process {c.name}'s order? [Y/N]: ").strip().lower()
    if process_or_not == 'y':
        if total_cost > c.budget:
            print("\n >>>> ERROR: Non-Sufficient Funds for the order to be processed")
            return 0
        else:
            process_order(s,c)
            remove_empty_values()



# PART 4: PROCESS CUSTOMER'S ORDER AND UPDATE STOCK.CSV
"""
Function that processes the customer's order
and calls function for updating stock.csv
"""
def process_order(s,c):
    for i in c.shopping_list:
        for j in s.stock:
            if i.product.name == j.product.name:
                if j.quantity < i.quantity:
                    print(f"\n >>>> ERROR: Not enough {i.product.name} in stock to fulfil the order.\n")
                    pass    
                else:
                    cost = i.quantity * i.product.price 
                    # Decrement the quantity of stock
                    j.quantity -= i.quantity
                    # Update the shop cash
                    s.cash += cost
                    # Decrement the customer's budget
                    c.budget -= cost
                    write_to_csv(i,j,s)
    print("\n -------------------------------------------------")
    print(f"\n {c.name}'s order has been successfully processed!")
    print(f" The balance on {c.name}'s account: €{c.budget}")
    print("\n -------------------------------------------------")


# UPDATE THE STOCK CSV                 
def write_to_csv(i,j,s):
    # Create a temporary file and fieldnames
    tempfile = NamedTemporaryFile(mode='w', delete=False)
    fields = ['Product', 'Cost', 'Quantity']
    try:
        with open(filename, 'r', newline='') as csvfile, tempfile:
            reader = csv.DictReader(csvfile, delimiter=',', lineterminator='\n', fieldnames=fields)
            writer = csv.DictWriter(tempfile, delimiter=',', lineterminator='\n', fieldnames=fields)
            for row in reader:
                # Check if Cost column is missing (to select the row with cash)
                # Adapted from https://stackoverflow.com/questions/34192705/python-how-to-check-if-cell-in-csv-file-is-empty
                if row['Cost'] in (None, ""):
                    row['Product'] = s.cash
                elif row['Product'] == i.product.name: 
                    row['Quantity'] = j.quantity
                    print('\n\t',row['Product'],'ordered')
                row = {'Product': row['Product'], 'Cost': row['Cost'], 'Quantity': row['Quantity']}
                writer.writerow(row)
        # Move the temporary file to stock.cv
        shutil.move(tempfile.name, filename)
    except FileNotFoundError:
        print(f" >>>> ERROR: cannot open {filename}")


"""
Function that removes empty values from the first line of stock.csv
(this deletes extra commas from the previous output beside cash in the csv)
"""
def remove_empty_values():        
    with open(filename, 'r', newline='') as csvfile:
        reader = csv.reader(csvfile, delimiter=',', lineterminator='\n')
        with open('../stock.temp.csv', 'w', newline='') as temp:
            writer = csv.writer(temp, delimiter=',', lineterminator='\n')
            for row in reader:
                new_row = [i for i in row if i]
                writer.writerow(new_row)
        shutil.move('../stock.temp.csv', filename)


    

# PART 4: MENU
def menu():
    print("\n PRESS\n\t-[1] to View shop\n\t-[2] to Place an order \n\t-[3] to Import customer's orders from a file\n\t-[0] to Exit")
    choice = input("\n Your choice: ").strip()
    return choice

choicemap = {
    '1': view_shop,
    '2': place_order,
    '3': import_order,
    '0': quit
}

choice = menu()
while choice != '0':
    if choice in choicemap:
        choicemap[choice] ()
    else:
        print(" Incorrect entry.")
    print("\n PRESS\n\t-[1] to View shop\n\t-[2] to Place an order \n\t-[3] to Import customer's orders from a file\n\t-[0] to Exit")
    choice = input("\n Your choice: ").strip()

File: Shop/C/shop.c
/* Olga Rozhdestvina
GMIT MULTI-PARADIGM-PROGRAMMING 2020
Shop Simulator in C 
Procedural Programming
*/

#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <ctype.h>
#include <errno.h>

struct Product
{
	char *name;
	double price;
};

struct ProductStock
{
	struct Product product;
	int quantity;
};

struct Shop
{
	double cash;
	struct ProductStock stock[20];
	int index;
};

struct Customer
{
	char *name;
	double budget;
	struct ProductStock shoppingList[10];
	int index;
};


// PART 1: STOCK SHOP

// Stock the shop from CSV using the above classes.
struct Shop createAndStockShop()
{
	FILE *fp;
	char *line = NULL;
	size_t len = 0;
	ssize_t read;
	char *filename = "../stock.csv"; 
	fp = fopen(filename, "r");
	if (fp == NULL)
	{
		fprintf(stderr, ">>>> ERROR: cannot open %s: \n", filename);
		exit(EXIT_FAILURE);
	}
	// Read the first line from stock.csv as initial cash of the shop.
	read = getline(&line, &len, fp);
	double initialCash = atof(line);
	struct Shop shop = {initialCash};

	// Stock up the shop.
	while ((read = getline(&line, &len, fp)) != -1)
	{
		char *n = strtok(line, ",");
		char *p = strtok(NULL, ",");
		char *q = strtok(NULL, ",");
		int quantity = atoi(q);
		double price = atof(p);
		char *name = malloc(sizeof(char) * 50);
		strcpy(name, n);

		struct Product product = {name, price};
		struct ProductStock stockItem = {product, quantity};
		shop.stock[shop.index++] = stockItem;
	}

	return shop;
}

// Display the shop
void printShop(struct Shop s)
{
	printf("\n *********************************************\n\n");
    printf("\t CASH AT THE SHOP: €%.2f\n", s.cash);
    printf("\n *********************************************\n");
    printf("\n Product\t\tPrice\t\tQty\n");
    printf(" _____________________________________________\n");
	for (int i = 0; i < s.index; i++)
	{
		printf("\n %s", s.stock[i].product.name);
        printf("\n\t\t\t%.2f\t\t%d\n", s.stock[i].product.price, s.stock[i].quantity);
        printf(" ---------------------------------------------\n");
	}
}


// Function that checks if product name is in the shop and returns struct product
struct Product getProduct(struct Shop s, char *pname)
{
	// check the shop for a product and return it
	struct Product p;
	for (int i = 0; i < s.index; i++)
	{
		// Adapted from https://stackoverflow.com/questions/30702039/how-to-do-a-case-insensitive-string-comparison
		if (strcasecmp(s.stock[i].product.name, pname) == 0)
		{
			p = s.stock[i].product;
		}
	}
	return p;
};

// Strip of extra spaces if tiped in
// Adapted from https://stackoverflow.com/questions/1488372/mimic-pythons-strip-function-in-c
char *strstrip(char *pname)
{
        size_t size;
        char *end;

        size = strlen(pname);

        if (!size)
                return pname;

        end = pname + size - 1;
        while (end >= pname && isspace(*end))
                end--;
        *(end + 1) = '\0';

        while (*pname && isspace(*pname))
                pname++;

        return pname;
}

// PART 2: LIVE MODE

// Create a shopping list from customer's input
void shoppingList(struct Shop *s, struct Customer *customer)
{
	char continueOrder;
	while(strcasecmp(&continueOrder, "N") != 0)
	{	
		char pn[20];
		float productQuantity;
		char *productName = malloc(sizeof(char) * 50);

		// Check if the product entered matches the shop stock.
		for (int i = 0; i < s->index; i++)
		{
				// Ask a customer to enter products and their quantity
				printf("\n Please enter product: ");
				scanf("\n %[^\n]%*c", &pn);
				fflush(stdin);

				printf(" Please enter quantity: ");
				scanf("%f", &productQuantity);
				fflush(stdin);

				// See if the product entered exists in the shop.
				strcpy(productName, pn);
				strstrip(pn);
				getProduct(*s, pn);
				double totalCost = productQuantity * getProduct(*s, pn).price;
				printf("\n >>>> The cost of it is €%.2f.\n", totalCost);

				// If the customer wants to add to the basket, append to the shopping list if yes.
				char addToBasket = printf("\n Would you like to add it to your basket? [Y/N]: ");
				scanf("%s", &addToBasket);
				fflush(stdin);
				if ((addToBasket == 'y') || (addToBasket == 'Y'))
					{	
						struct Product product = {getProduct(*s, pn).name, getProduct(*s, pn).price};
						struct ProductStock listItem = {product, productQuantity};
						customer->shoppingList[customer->index++] = listItem;
					}
				printf("\n Would you like to order something else? [Y/N]: ");
				scanf("%s", &continueOrder);
				fflush(stdin);
				if ((continueOrder == 'n') || (continueOrder == 'N')){
					break;
				}
				printf(" -------------------------------------------------\n");
		}
	}
}


/*
Function that collects data from a customer
and calls for function that creates a shopping list
*/
struct Customer placeOrder() 
{
	// Clear screen adapted from https://www.quora.com/What-is-the-function-of-a-system-cls-in-C-programming
	system("clear");
	struct Shop s = createAndStockShop();

	// Ask customer to enter his/her details
	char name[20];
	float budget;
	printf("Please enter your name: ");
	scanf("%s", &name);
	fflush(stdin);
	printf("Please enter your budget: ");
	scanf("%f", &budget);
	fflush(stdin);

	// Register customer's name and budget in the struct.
	char *customerName = malloc(sizeof(char) * 50);
	strcpy(customerName, name);
	strstrip(customerName);
	struct Customer customer = {customerName, budget};
	// Greet the customer (the first letter of the name capitalized)
	// Adapted from https://www.codevscolor.com/c-toupper-method-explanation-with-example
	printf("\n\t Hello, %c%s!\n", toupper(name[0]), name+1);
    printf(" +++++++++++++++++++++++++\n");
    printf("\n Choose from our products:\n\n");
	// Display available products
	for (int i = 0; i < s.index; i++)
	{
        printf(" %s\n", s.stock[i].product.name);
	}

	// Create a shopping list.
	shoppingList(&s,&customer);
	return customer;
}
	



// PART 3: READ CUSTOMER ORDER FROM CSV


// Load a customer's CSV file and create a shopping list
struct Customer readCustomer(char *filename)
{
	FILE *fp;
	char *line = NULL;
	size_t len = 0;
	ssize_t read;

	fp = fopen(filename, "r");
	if (fp == NULL)
		exit(EXIT_FAILURE);

	// Read the first line from customer file as customer's name and budget
	double customerLine = getline(&line, &len, fp);
	char *name = strtok(line, ",");
	char *budget = strtok(NULL, ",");
	char *customerName = malloc(sizeof(char) * 50);
	strcpy(customerName, name);
	double customerBudget = atof(budget);

	struct Shop s = createAndStockShop();
	struct Customer customer = {customerName, customerBudget};

	while ((read = getline(&line, &len, fp)) != -1)
	{
		char *pn = strtok(line, ",");
		char *qn = strtok(NULL, ",");
		int productQuantity = atoi(qn);
		char *productName = malloc(sizeof(char) * 50);
		strcpy(productName, pn);

		// Create a product and product stock with the information from the file
		struct Product product = {productName, getProduct(s, productName).price};
		struct ProductStock listItem = {product, productQuantity};
		// Increment the index and put the stock item in to the stock array
		customer.shoppingList[customer.index++] = listItem;
	}
	return customer;
}



// PART 4: PROCESS CUSTOMER'S ORDER AND UPDATE STOCK.CSV

// Update the stock.csv file.
// Followed from https://www.guru99.com/c-file-input-output.html
void updateShop(struct Shop s, struct Customer c)
{
	// Open the stock.csv in writing mode.
	FILE * fp;
	char *filename = "../stock.csv"; 
	fp = fopen(filename,"w+");
	if (fp == NULL)
	{
		fprintf(stderr, "Could not open %s: %s\n", filename, strerror(errno));
		exit(EXIT_FAILURE);
	}
	// Update cash.
	fprintf(fp,"%.2f\n", s.cash);

	// Update stock.
	for (int i=0; i< s.index; i++)
	{
		fprintf(fp, "%s,%.2f,%i\n",s.stock[i].product.name,s.stock[i].product.price,s.stock[i].quantity);	
	}
	fclose(fp);	
	return;
}


/* Function that processes the customer's order
and calls function for updating stock.csv 
*/
void processOrder(struct Shop *s, struct Customer *c)
{
	double cost = 0.0;

	// Check if we have each item on the order in stock
	for (int i = 0; i < c->index; i++)
	{
		for (int j = 0; j < s->index; j++)
		{
			if (strcmp(c->shoppingList[i].product.name, s->stock[j].product.name) == 0)
			{
				// Check if there is enough stock
				if (s->stock[j].quantity < c->shoppingList[i].quantity)
				{
					printf("\n >>>> ERROR: Not enough %s in stock to fulfill the order.\n", c->shoppingList[i].product.name);
					break;
				}
				else
				{
					cost = c->shoppingList[i].product.price * c->shoppingList[i].quantity;
					// Decrement the quantity of stock
					s->stock[j].quantity -= c->shoppingList[i].quantity;
					// Update the shop cash
					s->cash += c->shoppingList[i].product.price * c->shoppingList[i].quantity;
					// Decrement the customer's budget
					c->budget -= c->shoppingList[i].product.price * c->shoppingList[i].quantity;
					printf("\n \t%s ordered\n", c->shoppingList[i].product.name);			
				}
			}
		}
		updateShop(*s, *c);
	}
	printf("\n -------------------------------------------------\n");
    printf("\n %c%s's order has been successfully processed!", toupper(c->name[0]),c->name+1);
    printf("\n The balance on %c%s's account: €%.2f\n", toupper(c->name[0]),c->name+1, c->budget);
    printf("\n -------------------------------------------------\n");
}


/* Function that displays customer's credentials,
their structured shopping list,
calculates total cost of the order,
checks if the customer has enough funds to pay for the transaction
*/
void customersOrder(struct Customer c)
{
	system("clear");
	double totalCost = 0.0;
	struct Shop s = createAndStockShop();

	// Print the credentials.
	printf("\n *********************************************\n");
    printf("\n CUSTOMER: %c%s\n BUDGET:   €%.2f\n", toupper(c.name[0]), c.name+1, c.budget);
    printf("\n *********************************************\n");
	
	// Print the shopping list.
    printf("\n Product\t\t Price\t Qty\t Cost\n");
    printf("\n _____________________________________________\n");

	for (int i = 0; i < c.index; i++)
	{
		// Calculate the cost of each item
		double cost = c.shoppingList[i].quantity * c.shoppingList[i].product.price;
		if (c.shoppingList[i].quantity > 10) {
            printf(" Discounted\n %s", c.shoppingList[i].product.name);
            double discounted = cost - (cost * 0.1);
            printf("\n \t\t\t  %.2f \t  %d \t %.2f", c.shoppingList[i].product.price, c.shoppingList[i].quantity, discounted);
			printf("\n ---------------------------------------------\n\n");
			// add the line cost to the total
            totalCost += discounted;
		}
        else {
			printf("\n %s", c.shoppingList[i].product.name);
			printf("\n \t\t\t  %.2f \t  %d \t %.2f", c.shoppingList[i].product.price, c.shoppingList[i].quantity, cost);
			printf("\n ---------------------------------------------\n\n");
			// Update total cost of the order.
			totalCost += cost;
		}	
	}

	printf("\n Total\t\t\t\t\t %.2f", totalCost);
    printf("\n ---------------------------------------------\n");
	char processOrNot = printf("\n\n Would you like to process %c%s's order? [Y/N]: ", toupper(c.name[0]), c.name+1);
	scanf("%s", &processOrNot);
	fflush(stdin);

	// Check if customer has enough money to pay for the order.
    if ((processOrNot == 'y') || (processOrNot == 'Y'))
	{
        if (totalCost > c.budget)
		{
            printf("\n>>>> ERROR: Non-Sufficient Funds for the order to be processed\n");
            return;
		}
        else
		{	
			processOrder(&s,&c);
		}
	}
}


// Choose between 3 CVS files to read
void importOrder()
{
	int choice;
	char *filename;

	system("clear");
	fflush(stdin);
	printf("\n -----------------------------------------------------\n");
	printf("\n Please choose the file you would like to import.\n");
	printf("\n\n PRESS\n\t-[1] for Customer 1\n\t-[2] for Customer 2\n\t-[3] for Customer 3\n\t- or any other key to Return to the menu");
	printf("\n\n Your choice: ");
	scanf("%d", &choice);
	fflush(stdin);
	// normal order
	if (choice == 1)
	{
		filename = "../customer1.csv";
		struct Customer c = readCustomer(filename);
		customersOrder(c);
	}
	// not enough money
	else if (choice == 2)
	{
		filename = "../customer2.csv";
		struct Customer c = readCustomer(filename);
		customersOrder(c);
	}
	// not enough stock
	else if (choice == 3)
	{
		filename = "../customer3.csv";
		struct Customer c = readCustomer(filename);
		customersOrder(c);
	}	
}


// PART 5: MENU
// Adapted from https://stackoverflow.com/questions/15469064/interactive-menu-with-switch-statement
void menu(struct Shop s)
{
	printf("\n\t\tWelcome to Olga's Shop!\n+++++++++++++++++++++++++++++++++++++++++++++++++++++\n\n\n");
	int choice;
	fflush(stdin);
	do {
		// Menu options
		printf("\n PRESS\n\t-[1] to View shop\n\t-[2] to Place an order \n\t-[3] to Import customer's orders from a file\n\t-[0] to Exit");
		printf("\n\n Your choice: ");
		scanf("%d", &choice);

		switch (choice)
		{
			case 1:
					system("clear");
					struct Shop s = createAndStockShop();
					printShop(s);
					break;
			case 2:
					system("clear");
					struct Customer c = placeOrder();
					customersOrder(c);
					break;
			case 3:
					system("clear");
					importOrder(s);
					break;
			case 0:
					break;
			default:	
					printf("\n Incorrect entry.\n");
					break;
		}
	}	
	while (choice != 0);
}


// The main function that calls the menu.
int main(void)
{
	struct Shop shop = createAndStockShop();
	system("clear");
	// call the main screen
	menu(shop);
	return 0;
}